model/riscv_insts_zcmp.sail

```diff
@@ -0,0 +1,224 @@
+  0x4 <-> "ra",
+  0x5 <-> "ra" ^ sep() ^ "s0",
+  0x6 <-> "ra" ^ sep() ^ "s0-s1",
+  0x7 <-> "ra" ^ sep() ^ "s0-s2",
+  0x8 <-> "ra" ^ sep() ^ "s0-s3",
+  0x9 <-> "ra" ^ sep() ^ "s0-s4",
+  0xa <-> "ra" ^ sep() ^ "s0-s5",
+  0xb <-> "ra" ^ sep() ^ "s0-s6",
+  0xc <-> "ra" ^ sep() ^ "s0-s7",
+  0xd <-> "ra" ^ sep() ^ "s0-s8",
+  0xe <-> "ra" ^ sep() ^ "s0-s9",
+  0xf <-> "ra" ^ sep() ^ "s0-s11",
+  // Using X names instead of ABI names.
+  0x4 <-> "x1",
+  0x5 <-> "x1" ^ sep() ^ "x8",
+  0x6 <-> "x1" ^ sep() ^ "x8-x9",
+  0x7 <-> "x1" ^ sep() ^ "x8-x9" ^ sep() ^ "x18",
+  0x8 <-> "x1" ^ sep() ^ "x8-x9" ^ sep() ^ "x18-x19",
+  0x9 <-> "x1" ^ sep() ^ "x8-x9" ^ sep() ^ "x18-x20",
+  0xa <-> "x1" ^ sep() ^ "x8-x9" ^ sep() ^ "x18-x21",
+  0xb <-> "x1" ^ sep() ^ "x8-x9" ^ sep() ^ "x18-x22",
+  0xc <-> "x1" ^ sep() ^ "x8-x9" ^ sep() ^ "x18-x23",
+  0xd <-> "x1" ^ sep() ^ "x8-x9" ^ sep() ^ "x18-x24",
+  0xe <-> "x1" ^ sep() ^ "x8-x9" ^ sep() ^ "x18-x25",
+  0xf <-> "x1" ^ sep() ^ "x8-x9" ^ sep() ^ "x18-x27",
+}
+
+function get_stack_adj_base(rlist : bits(4)) -> int = {
+  if xlen == 32 then
+    match rlist[3 .. 2] {
+      0b01 => 16,
+      0b10 => 32,
+      0b11 => if rlist == 0b1111 then 64 else 48,
+      _    => 0
+    }
+  else {
+    match rlist[3 .. 1] {
+      0b010 => 16,
+      0b011 => 32,
+      0b100 => 48,
+      0b101 => 64,
+      0b110 => 80,
+      0b111 => if rlist == 0b1111 then 112 else 96,
+      _     => 0
+    }
+  }
+}
+
+mapping zcmp_assembly_mapping : (bits(4), bits(2), bool) <-> string = {
+  forwards (rlist, spimm54, is_negative) => {
+    let stack_adj_base = get_stack_adj_base(rlist);
+    let stack_adj = stack_adj_base + unsigned(spimm54) * 16;
+    let sign = if is_negative then "-" else "";
+
+    "{" ^ zcmp_assembly_rlist(rlist) ^ "}" ^ sep() ^ sign ^ dec_str(stack_adj)
+  },
+  backwards str_input => {
+    /* Backwards mapping for cm.push, cm.pop, cm.popret, and cm.popretz */
+    /* is not implemented due to the complexity of parsing these instructions. */
+    not_implemented("Backwards mapping for (cm.push/cm.pop/cm.popret/cm.popretz) is not implemented.");
+  },
+}
+
+function zcmp_regmask(rlist : bits(4)) -> bits(32) = {
+  var mask : bits(32) = zeros();
+
+  if rlist >=_u 0b0100 then mask[1] = bitone;
+
+  foreach (i from 5 to unsigned(rlist)) {
+    assert(i <= 15);
+    if i - 5 < 2
+    then mask[ 3 + i] = bitone
+    else mask[11 + i] = bitone
+  };
+
+  if rlist == zero_extend(0xf) then mask[27] = bitone;
+
+  mask
+}
+
+union clause ast = CM_PUSH : (bits(4), bits(2))
+
+mapping clause encdec_compressed = CM_PUSH(rlist, spimm)
+  <-> 0b101 @ 0b11000 @ rlist : bits(4) @ spimm : bits(2) @ 0b10
+  when currentlyEnabled(Ext_Zcmp) & rlist >=_u 0b0100
+
+function clause execute (CM_PUSH(rlist, spimm)) = {
+  let addr      = X(sp);
+  let stack_adj = negate_int(get_stack_adj_base(rlist) + unsigned(spimm) * 16);
+  let new_sp    = addr + to_bits(xlen, stack_adj);
```

Comment:
I think this can just be

```
let addr = addr - (get_stack_adj_base(rlist) + unsigned(spimm) * 16);
```

Or with the other suggestions

```
let addr = addr - get_stack_adj(rlist, spimm) * 16;
```